tests/llvm/test_helpers.py

```diff
@@ -220,3 +220,375 @@ def test_helper_printf(capfd, ir_argtype, format_spec, values_to_check):
+            bin_f.cuda_wrap_call(res)
+            res = res[0]
+
+        assert res.value == expected
+
+    @pytest.mark.llvm
+    @pytest.mark.parametrize('mode', ['CPU',
+                                  pytest.param('PTX', marks=pytest.mark.cuda)])
+    @pytest.mark.parametrize('ir_type,expected', [
+        (FLOAT_TYPE, 0),
+        (FLOAT_PTR_TYPE, 0),
+        (DOUBLE_TYPE, 0),
+        (DOUBLE_PTR_TYPE, 0),
+        (DOUBLE_VECTOR_TYPE, 0),
+        (DOUBLE_VECTOR_PTR_TYPE, 0),
+        (DOUBLE_MATRIX_TYPE, 0),
+        (DOUBLE_MATRIX_PTR_TYPE, 0),
+        (INT_TYPE, 0),
+        (INT_PTR_TYPE, 0),
+        (BOOL_TYPE, 1),
+        (BOOL_PTR_TYPE, 1),
+    ])
+    def test_helper_is_boolean(self, mode, ir_type, expected):
+        with pnlvm.LLVMBuilderContext() as ctx:
+            func_ty = ir.FunctionType(ir.VoidType(), [ir.IntType(32).as_pointer()])
+
+            custom_name = ctx.get_unique_name("is_boolean")
+            function = ir.Function(ctx.module, func_ty, name=custom_name)
+            out = function.args[0]
+            block = function.append_basic_block(name="entry")
+            builder = ir.IRBuilder(block)
+
+            variable = builder.load(builder.alloca(ir_type))
+            if pnlvm.helpers.is_boolean(variable):
+                builder.store(pnlvm.ir.IntType(32)(1), out)
+            else:
+                builder.store(pnlvm.ir.IntType(32)(0), out)
+
+            builder.ret_void()
+
+        bin_f = pnlvm.LLVMBinaryFunction.get(custom_name)
+        if mode == 'CPU':
+            res = ctypes.c_int32()
+
+            bin_f(ctypes.byref(res))
+        else:
+            res = np.array([False], dtype=np.bool_)
+            bin_f.cuda_wrap_call(res)
+            res = res[0]
+
+        assert res.value == expected
+
+@pytest.mark.llvm
+@pytest.mark.parametrize('mode', ['CPU',
+                                  pytest.param('PTX', marks=pytest.mark.cuda)])
+@pytest.mark.parametrize('op,var,expected', [
+    (pnlvm.helpers.tanh, 1.0, 0.7615941559557649),
+    (pnlvm.helpers.exp, 1.0, 2.718281828459045),
+])
+def test_helper_numerical(mode, op, var, expected):
+    with pnlvm.LLVMBuilderContext() as ctx:
+        func_ty = ir.FunctionType(ir.VoidType(), [ctx.float_ty.as_pointer()])
+
+        custom_name = ctx.get_unique_name("numerical")
+        function = ir.Function(ctx.module, func_ty, name=custom_name)
+        out = function.args[0]
+        block = function.append_basic_block(name="entry")
+        builder = ir.IRBuilder(block)
+
+        variable = ctx.float_ty(var)
+        result = op(ctx, builder, variable)
+        builder.store(result, out)
+
+        builder.ret_void()
+
+    bin_f = pnlvm.LLVMBinaryFunction.get(custom_name)
+    if mode == 'CPU':
+        res = ctypes.c_double()
+
+        bin_f(ctypes.byref(res))
+    else:
+        res = np.array([0.0], dtype=np.float)
+        bin_f.cuda_wrap_call(res)
+        res = res[0]
+
+    assert res.value == expected
```

Comment:
I'm surprised this passes. I'd expect it to need 'allclose', but let's keep it this way, just to see if/when it breaks.

Reply:
more precision can't be bad, i guess

¯\_(ツ)_/¯

Reply:
It can be slower. The ancient x87 fp operations use 80b representation of doubles internally, the newer (SSE+, or other than x86 arch) use 64b doubles.
I'm more worried about inconsistent results between platforms depending how the respective `libm` implements the functions, but if both github CI and travis are happy we're good.